packages/vincent-tools/vincent-tool-erc20-approval/src/lib/vincent-tool.ts

```diff
@@ -16,80 +16,69 @@ export const vincentTool = createVincentTool({
+    const { rpcUrl, spenderAddress, tokenAddress, tokenDecimals, tokenAmount } = toolParams;
 
-    const client = createPublicClient({
-      transport: http(rpcUrl),
-    });
+    const provider = new ethers.providers.JsonRpcProvider(rpcUrl);
 
     await checkNativeTokenBalance({
-      client,
-      pkpEthAddress: pkpEthAddress as `0x${string}`,
+      provider,
+      pkpEthAddress: ethAddress,
     });
 
     const currentAllowance = await getCurrentAllowance({
-      client,
-      tokenAddress: tokenAddress as `0x${string}`,
-      owner: pkpEthAddress as `0x${string}`,
-      spender: spenderAddress as `0x${string}`,
+      provider,
+      tokenAddress,
+      owner: ethAddress,
+      spender: spenderAddress,
     });
 
-    const requiredAmount = parseUnits(tokenAmount.toString(), tokenDecimals);
+    const requiredAmount = ethers.utils.parseUnits(tokenAmount.toString(), tokenDecimals);
 
     return succeed({
       allow: true,
-      existingApprovalSufficient: currentAllowance >= requiredAmount,
+      existingApprovalSufficient: currentAllowance >= requiredAmount.toBigInt(),
     });
   },
-  execute: async ({ toolParams }, { succeed }) => {
+  execute: async ({ toolParams }, { succeed, delegation: { delegatorPkpInfo } }) => {
     console.log('Executing ERC20 Approval Tool');
 
-    const {
-      rpcUrl,
-      chainId,
-      pkpEthAddress,
-      spenderAddress,
-      tokenAddress,
-      tokenDecimals,
-      tokenAmount,
-    } = toolParams;
-
-    const pkpInfo = await getPkpInfo({
-      pkpEthAddress: pkpEthAddress as `0x${string}`,
-    });
+    const { ethAddress, publicKey } = delegatorPkpInfo;
+    const { rpcUrl, chainId, spenderAddress, tokenAddress, tokenDecimals, tokenAmount } =
+      toolParams;
 
-    const client = createPublicClient({
-      transport: http(rpcUrl),
-    });
+    const provider = new ethers.providers.JsonRpcProvider(rpcUrl);
 
     await checkNativeTokenBalance({
-      client,
-      pkpEthAddress: pkpEthAddress as `0x${string}`,
+      provider,
+      pkpEthAddress: ethAddress,
     });
 
     const currentAllowance = await getCurrentAllowance({
-      client,
-      tokenAddress: tokenAddress as `0x${string}`,
-      owner: pkpEthAddress as `0x${string}`,
-      spender: spenderAddress as `0x${string}`,
+      provider,
+      tokenAddress,
+      owner: ethAddress,
+      spender: spenderAddress,
     });
+    const requiredAmount = ethers.utils
+      .parseUnits(tokenAmount.toString(), tokenDecimals)
+      .toBigInt();
 
-    const requiredAmount = parseUnits(tokenAmount.toString(), tokenDecimals);
+    console.log(
+      `currentAllowance: ${currentAllowance} >= requiredAmount: ${requiredAmount} (execute)`,
+    );
 
-    if (currentAllowance >= requiredAmount) {
-      console.log(
-        `currentAllowance: ${currentAllowance} >= requiredAmount: ${requiredAmount} (execute)`,
-      );
+    if (currentAllowance >= requiredAmount && requiredAmount > 0n) {
```

Comment:
This `if` seems opinionated. What if I want to reduce the allowance?
I would remove it entirely. Mostly because this tool could be isolated or included in another app with different goals and work just fine, I should not be restricted to only increase or erase the allowance

Reply:
:+1: seems reasonable to me. I made this clearer and enhanced the result of `precheck()`'s deny case slightly